staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go

```diff
@@ -162,34 +171,74 @@ func NodeV1alpha3(enabled bool) Option {
 	}
 }
 
+// KubeClient grants the plugin access to the API server. This is needed
+// for syncing ResourceSlice objects. It's the responsibility of the DRA driver
+// developer to ensure that this client has permission to read, write,
+// patch and list such objects. It also needs permission to read node objects.
+// Ideally, a validating admission policy should be used to limit write
+// access to ResourceSlices which belong to the node.
+func KubeClient(kubeClient kubernetes.Interface) Option {
+	return func(o *options) error {
+		o.kubeClient = kubeClient
+		return nil
+	}
+}
+
+// NodeName tells the plugin on which node it is running. This is needed for
+// syncing ResourceSlice objects.
+func NodeName(nodeName string) Option {
+	return func(o *options) error {
+		o.nodeName = nodeName
+		return nil
+	}
+}
+
 type options struct {
 	logger                     klog.Logger
 	grpcVerbosity              int
 	driverName                 string
+	nodeName                   string
 	draEndpoint                endpoint
 	draAddress                 string
 	pluginRegistrationEndpoint endpoint
 	unaryInterceptors          []grpc.UnaryServerInterceptor
 	streamInterceptors         []grpc.StreamServerInterceptor
+	kubeClient                 kubernetes.Interface
 
 	nodeV1alpha3 bool
 }
 
 // draPlugin combines the kubelet registration service and the DRA node plugin
 // service.
 type draPlugin struct {
-	registrar *nodeRegistrar
-	plugin    *grpcServer
+	ctx        context.Context // For new background activities that are started later.
```

Comment:
Ok, I mentioned this elswhere, but generally keeping context in a struct is [not recommended](https://go.dev/blog/context-and-structs). Is there a reason these background activities shouldn't just use a new context? Is there precedent for this elsewhere in K8s?

Reply:
In unit tests, the context given to the plugin enables per-test logging and cancellation.

In the kubelet it might do contextual logging, I would need to check to be sure. There is no separate cancellation.

There is at least one other precedent in K8s (event publishing). I'm gradually converting more code to use contexts instead of falling back to `context.Background` or `context.TODO` (only 3621 left...) and sometimes a context has to be stored in a struct. This is not a hard rule.

Reply:
If there are specific settings associated with the Plugin (like logging or cancellation), I think the best practice would be to make those configurable options in the plugin itself, and then have the plugin use those when it creates its own Context. That avoids stashing away an opaque context object which could contain anything.

Reply:
But that's the point of a context: code B that sits between "caller A" and "callee C" shouldn't need to know what A wants to pass to C. For cancellation, the plugin should accept a context and pass that on to other code.

Reply:
is an instance of the draPlugin struct long-lived or scoped to a single request?

Reply:
Exactly, what I am wondering is if we are reusing the stashed context for multiple, unrelated requests.

Reply:
The advantage of deriving all contexts in a test or program from the same root context is that cancellation propagates automatically. I'm fed up with investigating goroutine leaks because some custom shutdown method wasn't called.

Reply:
but... this isn't test-only code, right?

Reply:
> is an instance of the draPlugin struct long-lived or scoped to a single request?

The DRA plugin is long-lived.

>  I am wondering is if we are reusing the stashed context for multiple, unrelated requests.

We don't. If a request comes with its own context, then that context is used, not the background context of the plugin.

> but... this isn't test-only code, right?

Yes, but the same code runs in two different "contexts": as part of the program and as part of individual unit tests. The "must shutdown on cancellation" is more important for unit tests.

Reply:
There is exactly one place where `d.ctx` is used:
https://github.com/kubernetes/kubernetes/blob/5f6ae717e1d70e6e01ff8f36821e87dc28c37dad/staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go#L339-L341

So suppose `d.ctx` were removed. Which context should the background activity that is started in that function use?

It cannot be the one from `PublishResources` (the API which triggers publishing) because that is not connected to the cancellation handling that was set up for the plugin in `Start`. Also, `PublishResources` gets called more than once, so using it's context for some longer running background activitiy would be wrong.

@johnbelamaric's suggestion points towards starting with `context.Background` and then carefully adding whatever was passed to `Start`, plus connecting cancellation. That's more complicated than passing the context through the struct.

Reply:
`Start` could directly start publishing. But then we run that resource slice controller already before we know if it really should run and what to publish. To change that, we would need additional options for `Start` that we don't need right now.

Reply:
The context is basically a big bucket of "anything" and the conventions around context are such that callers passing a context do not expect it to live forever. Since we don't know what's in there, we could result in memory or other resource leaks as we hang on to references to objects that are long since unneeded. Even if that's not the case here, I am concerned about a precedent that could lead to mysterious and hard to bug issues later. Essentially, stashing contexts is semantically equivalent to global variables, in my mind, and carries all the same baggage.

Instead, I would prefer to see valid, understood options exposed as such, and included in the plugin API.

Reply:
> The context is basically a big bucket of "anything" and the conventions around context are such that callers passing a context do not expect it to live forever. 

In this case, the entire API is defined as "this context is being used for the background activity that you are starting". It shouldn't be a surprise that the context remains in use after `Start` returns - that's it's purpose, because it provides the cancellation signal for the background activity!

We do the same in other APIs which start long-running background activities, for example for controllers in kube-controller-manager.

I could rewrite `Start` so that it doesn't stash the context, but it would still remain in use.

Reply:
I see. So you are saying this isn't an ordinary context but rather one specifically and explicitly intended for this purpose. I guess I am ok with that.